File: eliHenderson/assignments/BreadthFirstSearch/lc117.js
// Problem Statement #

// Given a binary tree, connect each node with its level order successor. The last node of each level should point to a null node.

//  * this pattern follows the breadth first search pattern
//  *
//  * create a result array to hold the resulting tree
//  * create a queue
//  * push the root node into the queue
//  * while the queue is not empty
//  * create a temp array to hold the current level
//  * store the current size of the queue
//  * for loop from 0 to the current queue size
//  * hold the current node value
//  * current node is queue.shift, because we pushed into the queue
//  * push the current node into the current level temp array
//  * check if there are children
//  * if so push them into the queue
//  *
//  *  iterate of the current level array
//  *  if the node that follow the current node is null
//  *  current node will link to null
//  *  else
//  *  current node will link to the next node in the level (array)
//  *
//  * push the now linked, current level array into the results array
//  *
//  * return results array
//  *
//  * result = []
//  * queue = [root]
//  * while (queue not empty)
//  * current level = []
//   level size = queue length
//     for loop ( 0 - level size)
//       current node = queue.shift()
//       current level [current node]
//       if children push them into queue

//     for loop (0 - current level size)
//       if Node + 1 === null
//         node.next = null
//       else
//         node.next = node + 1

//     result.push(current level)

//   return result

//   Time: O(N)
//   Space: O(N)

//   edge:
//     what to do null vals
//   Assumption:
//     null values are valid

//   edge:
//     what to do with null arguments
//   Assumption
//     return null

class TreeNode {
  constructor(val) {
    this.val = val;
    this.left = null;
    this.right = null;
    this.next = null;
  }

  // level order traversal using 'next' pointer
  print_level_order() {
    console.log("Level order traversal using 'next' pointer: ");
    let nextLevelRoot = this;
    while (nextLevelRoot !== null) {
      let current = nextLevelRoot;
      nextLevelRoot = null;
      while (current != null) {
        process.stdout.write(`${current.val} `);
        if (nextLevelRoot === null) {
          if (current.left !== null) {
            nextLevelRoot = current.left;
          } else if (current.right !== null) {
            nextLevelRoot = current.right;
          }
        }
        current = current.next;
      }
      console.log();
    }
  }
}

const connect_level_order_siblings = function (root) {
  if (root === null) return null;
  //create results array
  const result = [];
  //create a queue
  const queue = [];
  //push in the root
  queue.push(root);

  //while queue is not empty
  while (queue.length) {
    //create temp level array
    let currLvl = [];
    //save current queue size
    let levelSize = queue.length;
    //loop from 0 - current level size
    for (let i = 0; i < levelSize; i++) {
      //store the current level
      let currNode = queue.shift();
      //push the current node into the current Level array
      currLvl.push(currNode);
      //push children into the queue if they exist
      currNode.left && queue.push(currNode.left);
      currNode.right && queue.push(currNode.right);
    }

    //loop over the current level array
    for (let j = 0; j < currLvl.length; j++) {
      //if the next node is not null
      if (currLvl[j + 1] !== null) {
        //set next node as current nodes sibling
        currLvl[j].next = currLvl[j + 1];
      } else {
        //set next to null
        currLvl[j].next = null;
      }
    }
    //push the temp current array into the results array
    result.push(currLvl);
  }

  //return results array
  return result;
};

const connect_level_order_siblings = (root) => {
  if (root === null) return null;
  //create results array
  const result = [];
  //create a queue
  const queue = [];
  //push in the root
  queue.push(root);

  //while queue is not empty
  while (queue.length) {
    //create temp level array
    let currLvl = [];
    //save current queue size
    let levelSize = queue.length;
    //loop from 0 - current level size
    for (let i = 0; i < levelSize; i++) {
      //store the current level
      let currNode = queue.shift();
      //push the current node into the current Level array
      currLvl.push(currNode);
      //push children into the queue if they exist
      currNode.left && queue.push(currNode.left);
      currNode.right && queue.push(currNode.right);
    }

    //loop over the current level array
    for (let j = 0; j < currLvl.length; j++) {
      //if the next node is not null
      if (currLvl[j + 1] !== null) {
        //set next node as current nodes sibling
        currLvl[j].next = currLvl[j + 1];
      } else {
        //set next to null
        currLvl[j].next = null;
      }
    }
    //push the temp current array into the results array
    result.push(currLvl);
  }

  //return results array
  return result;
};

var root = new TreeNode(12);
root.left = new TreeNode(7);
root.right = new TreeNode(1);
root.left.left = new TreeNode(9);
root.right.left = new TreeNode(10);
root.right.right = new TreeNode(5);
root.right.right.left = new TreeNode(null);
connect_level_order_siblings(root);

connect_level_order_siblings(null);

root.print_level_order();

File: taojieTan/assignments/treesBFS/lc117/lc117.py
# Problem Statement #

# Given a binary tree, connect each node with its level order successor. The last node of each level should point to a null node.

from __future__ import print_function

# input: binary tree
# output: an array with nodes
# edge cases: 1. empty tree: return []

'''
  Overview:
  we can apply BFS pattern using a queue to traverse through each node,
  while at each node, we connect the next pointer with the right node at the current level.

  Pseudo code:
  First initialize a queue with the root node
  Use while loop to check if the length of the queue is empty:
  inside the loop, initialize a length variable with the length of queue to represent number of nodes in current level.
  use for loop to traverse each node in the current level,
  pop the first node in the queue, check if current index + 1 is smaller than the for loop length,
  if yes, it means we have right node, we peak the first node in the queue and assign the next pointer with right node;
  if no, it means there isn't any level order successor, we assign the right pointer to None.
  
  Time complexity: O(N), space complexity: O(1) if not counting the queue. 
  
'''

class TreeNode:
  def __init__(self, val):
    self.val = val
    self.left, self.right, self.next = None, None, None

  # level order traversal using 'next' pointer
  def print_level_order(self):
    nextLevelRoot = self
    while nextLevelRoot:
      current = nextLevelRoot
      nextLevelRoot = None
      while current:
        print(str(current.val) + " ", end='')
        if not nextLevelRoot:
          if current.left:
            nextLevelRoot = current.left
          elif current.right:
            nextLevelRoot = current.right
        current = current.next
      print()


def connect_level_order_siblings(root):
  if root is None:
    return []

  queue = [root]

  while len(queue):
    num_of_nodes = len(queue)
    for i in range(num_of_nodes):
      node = queue.pop(0)
      if i + 1 < num_of_nodes:
        successor = queue[0]
        node.next = successor
      else:
        node.next = None   

      if node.left:
        queue.append(node.left)
      if node.right:
        queue.append(node.right)

  return root

def main():
  root = TreeNode(12)
  root.left = TreeNode(7)
  root.right = TreeNode(1)
  root.left.left = TreeNode(9)
  root.right.left = TreeNode(10)
  root.right.right = TreeNode(5)
  connect_level_order_siblings(root)

  print("Level order traversal using 'next' pointer: ")
  root.print_level_order()

'''
        12 
      7    1     
    9    10  5
                                12 -> None
               => return      7 -> 1 -> None
                            9 -> 10->5 -> None
'''

main()
